core/domain/user_services.py

```diff
@@ -2287,3 +2291,291 @@ def get_dashboard_stats(user_id):
         'num_ratings': num_ratings,
         'average_ratings': average_ratings
     }
+
+
+def _get_checkpoints_in_order(init_state_name, states):
+    """Returns the checkpoints of an exploration in sequential order by a
+    BFS traversal.
+
+    Args:
+        init_state_name: str. The name of the first state of the exploration.
+        states: dict(state). All states of the exploration.
+
+    Returns:
+        list(str). List of all checkpoints of the exploration in sequential
+        order.
+    """
+    queue = []
+    queue.append(init_state_name)
+    checkpoint_state_names = []
+    visited_state_names = []
+    while len(queue) > 0:
+        current_state_name = queue.pop()
+        if current_state_name not in visited_state_names:
+            visited_state_names.append(current_state_name)
+            current_state = states[current_state_name]
+            if (current_state.card_is_checkpoint and (
+                current_state_name not in checkpoint_state_names)):
+                checkpoint_state_names.append(current_state_name)
+            for answer_group in current_state.interaction.answer_groups:
+                queue.append(answer_group.outcome.dest)
+
+    return checkpoint_state_names
+
+
+def _get_most_distant_reached_checkpoint_in_current_exploration(
+    checkpoints_in_current_exploration,
+    checkpoints_in_older_exploration,
+    most_distant_reached_checkpoint_state_name_in_older_exploration
+):
+    """Returns the most distant reached checkpoint in current exploration after
+    comparing current exploration with older exploration.
+
+    Args:
+        checkpoints_in_current_exploration: list(str). The checkpoints of
+            current exploration in sequential order.
+        checkpoints_in_older_exploration: list(str). The checkpoints
+            of older exploration in sequential order.
+        most_distant_reached_checkpoint_state_name_in_older_exploration: str.
+            The state name of the most distant reached checkpoint in the older
+            exploration.
+
+    Returns:
+        str or None. The most distant checkpoint in current exploration or
+        None if most distant reached checkpoint of older exploration is not
+        present in current exploration.
+    """
+    # Index of the most_distant_reached_checkpoint in the older exploration.
+    mdrc_index = (
+        checkpoints_in_older_exploration.index(
+            most_distant_reached_checkpoint_state_name_in_older_exploration))
+
+    # Loop through checkpoints of furthest_reached_exploration backwards until
+    # a checkpoint is found that exists in current_exploration too.
+    while mdrc_index >= 0:
+        if checkpoints_in_older_exploration[mdrc_index] in (
+            checkpoints_in_current_exploration):
+            return checkpoints_in_older_exploration[mdrc_index]
+        mdrc_index = mdrc_index - 1
+
+    return None
+
+
+def update_learner_checkpoint_progress(
```

Comment:
This function is a bit long, would it be possible to split it up a bit?

Reply:
I have shortened it up by some formatting changes you suggested. It basically covers all conditions which might arise while updating learner's checkpoint progress (that's why a lot of If and Else) and it would be complicated to split this into parts.